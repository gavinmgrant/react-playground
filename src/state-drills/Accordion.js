import React from "react";

class Accordion extends React.Component {
    static defaultProps = { sections: [], 
    };

    state = { currentIndex: null,
    };

    handleButtonClick(index) {
        this.setState({ currentIndex: index })
    }

    renderSection(section, index, currentIndex) {
        return (
            <li className='Accordion-section' key={index}>
                <button type='button'
                    onClick={() => this.handleButtonClick(index)}>
                    {section.title}
                </button>
                {(currentIndex === index) && <p>{section.content}</p>}
            </li>
        ) 
    }

    render() {
        const { currentIndex } = this.state
        const {sections } = this.props
        return (
            <ul>
                {sections.map((section, index) =>
                    this.renderSection(section, index, currentIndex)
                )}
            </ul>
        )
    }
}

export default Accordion;