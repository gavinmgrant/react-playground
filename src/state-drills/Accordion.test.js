import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Accordion from './Accordion';

// array of objects, each with a title and content
const sections = [
    {
        title: 'Section 1',
        content: 'Lorem ipsum dolor sit amet consectetur adipisicing elit.',
    },
    {
        title: 'Section 2',
        content: 'Cupiditate tenetur aliquam necessitatibus id distinctio quas nihil ipsam nisi modi!',
    },
    {
        title: 'Section 3',
        content: 'Animi amet cumque sint cupiditate officia ab voluptatibus libero optio et?',
    },
];;

describe(`Accordion Component`, () => {
    it('renders without errors', () => {
        const li = document.createElement('li');
        ReactDOM.render(<Accordion />, li);
        ReactDOM.unmountComponentAtNode(li);
    })

    it('renders an empty li when sections prop not supplied', () => {
        const wrapper = shallow(<Accordion />)
        expect(toJson(wrapper)).toMatchSnapshot()
    })

    it('renders no sections as active by default', () => {
        const wrapper = shallow(<Accordion sections={sections} />)
        expect(toJson(wrapper)).toMatchSnapshot()
    })

    it('opens a clicked section', () => {
        const wrapper = shallow(<Accordion sections={sections} />)
        wrapper.find('button').at(1).simulate('click')
        expect(toJson(wrapper)).toMatchSnapshot()
    })

    it('only opens the last section when multiple sections have been clicked', () => {
        const wrapper = shallow(<Accordion sections={sections} />)
        wrapper.find('button').at(1).simulate('click')
        wrapper.find('button').at(2).simulate('click')
        expect(toJson(wrapper)).toMatchSnapshot()
    })
})
