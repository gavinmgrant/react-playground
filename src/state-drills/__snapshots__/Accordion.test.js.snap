// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion Component only opens the last section when multiple sections have been clicked 1`] = `
<ul>
  <li
    className="Accordion-section"
    key="0"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 1
    </button>
  </li>
  <li
    className="Accordion-section"
    key="1"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 2
    </button>
  </li>
  <li
    className="Accordion-section"
    key="2"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 3
    </button>
    <p>
      Animi amet cumque sint cupiditate officia ab voluptatibus libero optio et?
    </p>
  </li>
</ul>
`;

exports[`Accordion Component opens a clicked section 1`] = `
<ul>
  <li
    className="Accordion-section"
    key="0"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 1
    </button>
  </li>
  <li
    className="Accordion-section"
    key="1"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 2
    </button>
    <p>
      Cupiditate tenetur aliquam necessitatibus id distinctio quas nihil ipsam nisi modi!
    </p>
  </li>
  <li
    className="Accordion-section"
    key="2"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 3
    </button>
  </li>
</ul>
`;

exports[`Accordion Component renders an empty li when sections prop not supplied 1`] = `<ul />`;

exports[`Accordion Component renders no sections as active by default 1`] = `
<ul>
  <li
    className="Accordion-section"
    key="0"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 1
    </button>
  </li>
  <li
    className="Accordion-section"
    key="1"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 2
    </button>
  </li>
  <li
    className="Accordion-section"
    key="2"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 3
    </button>
  </li>
</ul>
`;
